### PR TITLE
Generic docker improvements

### LINKED ISSUE #1954: Docker volume support for storing sqlite3 database
#### Is your feature request related to a problem?
When I want to start using Dex IdP docker image, the easiest storage driver to configure is the sqlite3. But the only writable directories for user `1001` are `/tmp` and `/var/tmp`. I'd like to store sqlite3 datafile to docker volume to keep changes persistent between multiple Dex container incarnations.

Unfortunately, Docker doesn't allow to set volume ownership/permissions explicitly, but it inherits ownership from the directory, where the volume is being mounted.

#### Describe the solution you'd like to see
Create a writeable directory in Dockerfile, optionally assigning it a mount point using `VOLUME` instruction.

```dockerfile
RUN mkdir /data && chown -R 1001 /data
VOLUME /data

# existing line
USER 1001:1001
```

User can run container from image like:
```
docker run --rm -v /path/to/dex.yaml:/dex.yaml -v dex-data:/data dexidp/dex:v2.27.0 serve /dex.yaml
```
(assuming the dex.yaml points sqlite3 storage file to `/data/dex.db`)

Any changes (users, clients added via gRPC) will be kept in `dex-data` volume.

#### Describe alternatives you've considered
There are not many alternatives, except for building own image based on `dexidp/dex` image that contains this writable directory. Of course, users could also use another storage backend (like Postgres DB), but it is a more complex solution that could deter faint-hearted souls.

### PR DESCRIPTION
**Pull request by sagikazarmark**
#### Overview

Generic docker improvements

#### What this PR does / why we need it

This PR contains a number of improvements for the container image build:

- Improve order of instructions to better utilize the cache
- Copy every source file to the final image from the builder image
- Add an alias to the builder image
- Update base images
- Add a `/var/dex` data directory (closes #1954)
- Relocate module files to `/usr/local/src/dex` (see #1863, cc @heidemn-faro)

#### Special notes for your reviewer

These changes mostly affect the image building process.

#### Does this PR introduce a user-facing change?

```release-note
`/var/dex` can be used as a data directory for Dex
Go module files are relocated to `/usr/local/src/dex`
```

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -44,6 +43,8 @@ COPY --from=0 /go/bin/dex /usr/local/bin/dex
 
 USER 1001:1001
 
+WORKDIR /
```
Comment (mouchar) [inline Dockerfile:42]: `WORKDIR /` is completely no-op, because you already are in the `/` directory

Comment (mouchar) [reply Dockerfile:42]: You also have a typo in commit message:
```patch
-chore(docke)r: move workdir down in the final image
+chore(docker): move workdir down in the final image
```

**Review by mouchar — COMMENTED** [review]
You should mention writeable `/var/dex` in README.

File: Dockerfile (hunk 1 above)
Comment (sagikazarmark) [reply Dockerfile]: Thanks!

**Comment by mouchar** [comment]
LGTM

**Review by bonifaido — APPROVED** [review]
📦 👍

**Review by justaugustus — APPROVED** [review]

**Pull request merged**